### PR TITLE
Revert option override

### PR DESCRIPTION
**Pull request by jpakkane**
Since nobody stepped up to the plate to fix this, reverting the original commit.

This turned out to be a good thing since it had some questionable choices. For example there were compile option overrides for build, host _and_ target, even though there are _no_ compilation options for target. The correct way to do this is to get the options via `c_args` and `cross_c_args` et al.

For the curious among you, reverting this one commit took roughly four hours of uninterrupted work.

The commits may need rebasing/squashing. Let's see what CI has to say about this first.

### CODE REVIEW

File: mesonbuild/interpreter.py [hunk 1]
```diff
@@ -2470,9 +2465,11 @@ def do_subproject(self, dirname, kwargs):
         return self.subprojects[dirname]
 
     def get_option_internal(self, optname):
+        # Some base options are not defined in some environments, return the
+        # default value from compilers.base_options in that case.
         for d in chain(
-                [self.coredata.base_options, compilers.base_options, self.coredata.builtins],
-                self.coredata.get_all_compiler_options()):
+                  [self.coredata.base_options, compilers.base_options,
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[Flake8]__ [E126] continuation line over-indented for hanging indent ([link](https://sider.review/gh/repos/19784232/pulls/5253#issue-3978663))

<sub>posted by Sider</sub>

File: mesonbuild/interpreter.py [hunk 2]
```diff
@@ -2470,9 +2465,11 @@ def do_subproject(self, dirname, kwargs):
...
+                  self.coredata.builtins, self.coredata.compiler_options]):
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[Flake8]__ [E128] continuation line under-indented for visual indent ([link](https://sider.review/gh/repos/19784232/pulls/5253#issue-3978664))

<sub>posted by Sider</sub>

**Comment by dcbaker** [comment]
Which issue is this fixing? All of the issues opened for 0.50.1 have proposed patches already AFAICT

**Comment by jpakkane** [comment]
#5097

**Comment by dcbaker** [comment]
I need to think about this a bit, I thought @Ericson2314 was looking at this or I would have done it :/

I'm not entirely convinced that a revert is necessary or correct. Just because a generic data structure has a field that *can* be used for target properties but is None in most cases isn't a bug, nor is it a bad design choice.

In this case it seems like we really want to specialize that data structure. Would you give me a couple of hours to prototype the other solution and see what it would look like before we merge this? If I don't have anything today consider the request fulfilled.

**Comment by jpakkane** [comment]
> Just because a generic data structure has a field that can be used for target properties but is None in most cases isn't a bug, nor is it a bad design choice.

But in this case it is _always_ none. There are no `c_args` for the target machine. Only for build and host.

> Would you give me a couple of hours to prototype the other solution and see what it would look like before we merge this

We don't really need to merge this until the other MRs are merged. Though this may conflict with the other cross option one.

**Comment by dcbaker** [comment]
In this case, but that same datastructure is used to store the data out of the target_machine of the cross file. We could certainly add a specialization of that class that accessing target is an error, but I think keeping the interface simple and consistent is a maintainability benefit

**Comment by dcbaker** [comment]
btw, your test cases pass on master right now, I cherry picked them into a branch to start looking at the problem.

**Comment by dcbaker** [comment]
drrrrr, I can read, they skip

**Comment by Ericson2314** [comment]
I was planning on doing fix that issue. I had started with the other 0.50 regressions; this was the last one.

And, I think this specific path a very bad direction for `master`. Doing `native` vs `cross` will always end up with cross being a buggy second class citizen for Meson. `build` vs `host` vs `target` is the only way that's maintainable. Maybe target is unused now, but I plan on using it, and by that criterion `[target_machine]` should have never been added to the cross file---meson don't use it except to pass it to `meson.build` so it don't need built-in support.

**Comment by jpakkane** [comment]
I did try fixing the issue first but that turned out to be more difficult than expected. The place where the host and target options are set does not have _any_ connection to places where they are defined. You'd either need to pass everything down to the place where they are created or modify the resulting objects once they get higher up in the call stack. That just plain has never worked. In addition if you used `meson configure` it _would_ list the correct C standard version, but the overrider just threw it away.

This approach raises a lot of questions and complicate the UX for no advantage. For example the system allows for overriding compiler args etc for host and for target. Where would those flags be stored? How would they be presented to the user? Having `c_args`, `cross_c_args` (or `cross_host_c_args`) and `cross_target_c_args` is just plain bad UI because of option explosion.

The crux here is that there are three machine types but only two compiler types. Trying to force two to be three or three to be two is confusing and leads to weird and difficult bugs like this one. Keeping the two explicit and separate makes the code easier to understand because there are no magic adaptor classes in the middle.

> Maybe target is unused now, but I planned on using it

For what, specifically?

> by that criterion [target_machine] should have never been added to the cross file---meson don't use it except to pass it to meson.build so it don't need built-in support.

On the contrary, things such as cross compilers _need_ that info and target machine is the standard way of providing it. It is added and supported for very good reasons.

And nothing about this is not to say that something like this could not be brought back at some point.

**Comment by Ericson2314** [comment]
@jpakkane Thanks for the quick response. And I'm sorry I responded rather curtly. The changes you reverted were part of my months long sago culminating in https://github.com/mesonbuild/meson/pull/4010, which otherwise is so close, and I was scared to seeing it all rolled back at the last minute.

I'm going to start the the fix I would have done for this issue to double check my reasoning, and then respond to you more fully, but for now let me respond in brief.

>  >Maybe target is unused now, but I planned on using it
>
> For what, specifically?

For compilers' build systems. I dream of rewriting GCC's with Meson. The compiler itself, various boilderplate gen tools, and `libgcc` and fiends would all be subprojects. The catch is those are all built for different platforms.
 - The boilerplate generators are built for the build platform, and would be `native: true` subprojects.
 - The compiler itself is built for the host platform, and would be a `native: false` subproject.
 - `libgcc` however is built for the target platform, and would be a `native: ???` subproject. 

I first raised these plans (I think) in https://github.com/mesonbuild/meson/issues/3972 .

**Comment by jpakkane** [comment]
The fact is that 0.50.0 has been broken w.r.t this (and other things) for far too long. We _will_ get 0.50.1 ready for release this weekend. Any issue that is not fully fixed _will_ get reverted. There will be _no_ exceptions!

**Comment by Ericson2314** [comment]
@jpakkane That's fair, and I'm sorry I didn't attack these regressions with more energy out the door. I fixed the test for my other PR (there was a git snafu but when I get home I'll use that reflog to fix).

**Comment by Akaricchi** [comment]
>libgcc however is built for the target platform, and would be a native: ??? subproject.

In #4963 I proposed to eventually deprecate `native:` in favor of `machine:`, so you can write e.g. `machine : target_machine` there instead.

**Comment by jpakkane** [comment]
Thinking about this more it seems that this is just one instance of many where people would want to use more than one (or two) compilers in a single project. In addition there are:

- Android projects that build a native component, one for arm32, one for arm64, one for Mips, one for x86 (?) etc.
- Embedded projects that consist of more than one processor, say a board that has 10 different microcontrollers that need firmware where each one requires its own special snowflake toolchain.
- The GCC bootstrap thing that requires both a host and a target compiler

Supporting these all in the build files means supporting an _arbitrary_ number of compilers in a single build directory. This is completely infeasible and breaks the UX in so many ways it's not even funny.

An alternative approach would be to keep the current native/cross separation and instead come up with a way to run multiple separate projects at the same time with different settings. This would be some sort of an agglomerative build thingy. For the Android case you'd write an agglomeration file that would look something like this (pseudo-YAML only used for convenience):

```
- project
  - cross-file: android-arm32.cross
  - source_dir: projectdir
  - options: ['-Dbuild_java_parts=false']
- project:
  - cross-file: android-x86.cross
  - source_dir: projectdir
  - options: ['-Dbuild_java_parts=true']
[etc for the remaining platforms]
```

Meson would then create a superproject build dir with a Ninja file that merely recurses in the parts. Thus `ninja all` would become the commands `ninja -C arm-builddir && ninja -C x86-builddir && ..`.

For the GCC case you'd have a file looking like this:

```
-project
  - cross_file: hostcompiler.cross
  - source_dir: 'gcc-sources'
- project
  - cross: targetcompiler.cross
  - source_dir: 'libgcc-sources'
```

This would keep the toolchain selection and configuration complexity away from build definition files, where it should not be because it complicates things.

**Comment by nirbheek** [comment]
Btw, have you already looked at https://github.com/mesonbuild/meson/pull/5255 to see if it's correct enough for this stable release? This PR is a very large change, and I'm afraid that it might cause new regressions.

**Comment by jpakkane** [comment]
I'll review it later today.

File: mesonbuild/interpreter.py [hunk 3]
```diff
@@ -2489,9 +2484,11 @@ def do_subproject(self, dirname, kwargs):
...
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[Flake8]__ [E128] continuation line under-indented for visual indent ([link](https://sider.review/gh/repos/19784232/pulls/5253#issue-4059267))

<sub>posted by Sider</sub>

File: mesonbuild/interpreter.py [hunk 4]
```diff
@@ -2489,9 +2484,11 @@ def do_subproject(self, dirname, kwargs):
         return self.subprojects[dirname]
 
     def get_option_internal(self, optname):
+        # Some base options are not defined in some environments, return the
+        # default value from compilers.base_options in that case.
         for d in chain(
-                [self.coredata.base_options, compilers.base_options, self.coredata.builtins],
-                self.coredata.get_all_compiler_options()):
+                  [self.coredata.base_options, compilers.base_options,
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[Flake8]__ [E126] continuation line over-indented for hanging indent ([link](https://sider.review/gh/repos/19784232/pulls/5253#issue-4059266))

<sub>posted by Sider</sub>

**Pull request closed without merging**